### PR TITLE
Tracing scheduling process

### PR DESCRIPTION
**Pull request by CtfGo**
升级IRSchedule类，支持trace调优过程，涉及改动包括两部分：
1. 采用`接口和实现分离`的思路，将ScheduleHelper重命名为ScheduleImpl，并且将原先IRSchedule中原语的实现都迁移到该类中，IRSchedule作为外部接口，内部调用ScheduleImpl的实现。
2. IRSchedule中新增SchedulleDesc(`trace_`)数据成员，IRSchdule的每个原语接口被调用时会向`trace_`追加一条ScheduleDesc::Step结构，表明该原语的类型，输入、输出、属性等信息，从而支持重放。

### CODE REVIEW

File: cinn/ir/ir_schedule.cc [hunk 1]
```diff
@@ -1374,7 +1427,196 @@ void IRSchedule::CopyTransformAndLoopInfo(const Expr& block, const Expr& block_t
   VLOG(3) << "res is : " << res;
   std::vector<Expr> all_loops = this->GetLoops(block);
   CHECK(!all_loops.empty());
-  helper_.Replace(all_loops[0], res);
+  this->Replace(all_loops[0], res);
+}
+
+IRSchedule::IRSchedule() {}
+
+IRSchedule::IRSchedule(IRSchedule&& other) : impl_(std::move(other.impl_)), trace_(other.trace_) {}
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.cc]: Does `other.trace_` need to be moved?

File: cinn/optim/ir_copy.h [hunk 2]
```diff
@@ -18,10 +18,14 @@
 #include <vector>
 
 #include "cinn/ir/ir.h"
-#include "cinn/ir/ir_schedule.h"
 #include "cinn/ir/lowered_func.h"
 
 namespace cinn {
+
+namespace ir {
+class ModuleExpr;
```
Comment (zhhsplendid) [inline cinn/optim/ir_copy.h:26]: What's the purpose for changes in this file?

File: cinn/optim/ir_copy.cc [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include "cinn/common/ir_util.h"
 #include "cinn/ir/ir_mutator.h"
 #include "cinn/ir/ir_printer.h"
+#include "cinn/ir/ir_schedule.h"
```
Comment (zhhsplendid) [inline cinn/optim/ir_copy.cc:26]: What's the purpose for changes in this file?

**Review by zhhsplendid — COMMENTED** [review]
Overall LGTM, please answer my three doubts above.

File: cinn/optim/ir_copy.h (hunk 2 above)
Comment (CtfGo) [reply cinn/optim/ir_copy.h:26]: to skip a compilation error caused by the conflict between the `ginac` and `protobuf` library, which is triggered by `ir_schedule.h` when it adds the `schedule_desc.h` header that has a  protobuf-generated header `schedule_desc.pb.h`

File: cinn/optim/ir_copy.cc (hunk 3 above)
Comment (CtfGo) [reply cinn/optim/ir_copy.cc:26]: same as explanation  in `ir_copy.h`, this is part of the solution

File: cinn/ir/ir_schedule.cc (hunk 1 above)
Comment (CtfGo) [reply cinn/ir/ir_schedule.cc]: not necessary, but I will add this one to be more stable

File: cinn/optim/ir_copy.h (hunk 2 above)
Comment (zhhsplendid) [reply cinn/optim/ir_copy.h:26]: Suggest to find other way in the future

**Review by zhhsplendid — APPROVED** [review]

**Pull request merged**